src/what-unsafe-does.md

```diff
@@ -16,16 +16,19 @@ to your program. You definitely *should not* invoke Undefined Behavior.
 Unlike C, Undefined Behavior is pretty limited in scope in Rust. All the core
 language cares about is preventing the following things:
 
-* Dereferencing null, dangling, or unaligned pointers
+* Dereferencing null, dangling, or unaligned references or raw pointers
+* Performing out-of-bounds arithmetic for the computation of a struct/tuple
+  field address
 * Reading [uninitialized memory][]
 * Breaking the [pointer aliasing rules][]
-* Producing invalid primitive values:
-    * dangling/null references
+* Producing/obtaining invalid primitive values:
+    * dangling/null/unaligned references
     * null `fn` pointers
     * a `bool` that isn't 0 or 1
     * an undefined `enum` discriminant
     * a `char` outside the ranges [0x0, 0xD7FF] and [0xE000, 0x10FFFF]
-    * A non-utf8 `str`
+    * a non-utf8 `str`
+    * a compound type (`enum`/`struct`/array/tuple) with an invalid field
```

Comment:
this doesn’t seem to add anything?

Reply:
Previously it said "produce an primitive value". But it is also UB to do
```rust
let b: (bool, u8) = mem::uninitialized();
```
even though this does not really produce any primitive value. I think it is important to stress that we proceed recursively in compound data types. I am happy for suggestions for how to word this better.

Reply:
It produces a `bool` though?

Reply:
I think this is worth stating explicitly. IMO the previous statement could be interpreted as if creating a compound value would never be a problem, as if wrapping a type in a struct would "hide" the inner invariant -- which I think is a position someone might have. So we should be explicit about this.

Reply:
I agree, but they probably shouldn't go in a list about primitive values. Instead, change the heading of the list:

> Producing invalid primitive values (either alone or as a field of a compound type):
> * foo..
> * bar..

Reply:
yep, that seems a lot better

Reply:
Ah, I meant to change the header but forgot.